File: src/kokoro_link/infrastructure/prompt/initial_relationship.py
"""Prompt renderer for initial relationship seed."""

from __future__ import annotations

from kokoro_link.domain.entities.character_operator_relationship_seed import (
    CharacterOperatorRelationshipSeed,
)


def render_initial_relationship_seed_lines(
    seed: CharacterOperatorRelationshipSeed | None,
    *,
    include_address: bool = True,
    include_proactive_permission: bool = True,
) -> list[str]:
    """Render the relationship-seed context block.

    ``include_address`` controls the two address lines (稱呼使用者 /
    使用者怎麼稱呼你). The main chat prompt sets it ``False`` because the
    bidirectional address resolver now owns those terms (so the seed,
    learned persona, and global profile don't render three competing
    名字). Auxiliary prompts and the player-facing persona-snapshot API
    keep the default ``True`` so their behaviour and contract are
    unchanged. ``include_proactive_permission`` controls the cold-start
    authorization lines. Once the user has already started interacting,
    proactive permission is governed by the character's live proactive
    setting instead; carrying the original cold-start denial forward would
    make an old setup silently veto later proactive evaluation.
    """
    if seed is None or seed.is_empty:
        return []
    lines = [
        "使用者創角時確認的起始關係設定：",
        "- 這些內容只能用於稱謂、語氣距離與互動邊界；不可改寫成已發生過的系統內記憶。",
        "- 未提供的共同經歷不得補完；資訊不足時請自然詢問，而不是假裝知道。",
    ]
    _append(lines, "關係", seed.relationship_label)
    _append(lines, "可知道的背景", seed.known_context)
    _append(lines, "居住安排", seed.living_arrangement)
    if include_address:
        _append(lines, "稱呼使用者", seed.user_address_name)
        _append(lines, "使用者怎麼稱呼你", seed.character_address_name)
    _append(lines, "語氣距離", seed.tone_distance)
    _append(lines, "熟悉度邊界", seed.familiarity_boundary)
    _append(lines, "行程中的使用者參與程度", _schedule_policy_text(seed))
    if include_proactive_permission:
        if seed.proactive_permission:
            _append(lines, "主動訊息授權", "使用者允許創角後主動找她／他，但必須遵守頻率與邊界。")
            _append(lines, "主動訊息頻率或時機", seed.proactive_cadence_hint)
        else:
            lines.append("- 主動訊息授權：沒有明確授權；不要把起始關係解讀成可直接打擾。")
    _append(lines, "使用者補充", seed.user_profile_notes)
    return lines


def render_arc_planner_relationship_lines(
    seed: CharacterOperatorRelationshipSeed | None,
) -> list[str]:
    """Facts a *planner* needs to place the player inside a story (OP1-A).

    Deliberately a narrower projection than
    :func:`render_initial_relationship_seed_lines`: the arc planner is
    not conducting the relationship, it is deciding — per beat — whether
    the player is absent, present or central, and what the player's
    dramatic place in that scene is. For that it needs to know **how the
    player is addressed** and **how close the two of them stand**;
    everything the chat block adds on top (proactive authorisation,
    schedule-involvement policy, invented-memory guards) belongs to a
    conversation that is happening now, not to a plan for the next three
    weeks.

    Returns ``[]`` for a missing / empty seed *and* for a seed whose
    every relevant field is blank — the caller renders no heading at all
    in that case, because a heading over nothing is a hole the model
    will feel obliged to fill.
    """
    if seed is None or seed.is_empty:
        return []
    lines: list[str] = []
    _append(lines, "角色怎麼稱呼玩家", seed.user_address_name)
    _append(lines, "玩家怎麼稱呼角色", seed.character_address_name)
    _append(lines, "關係", seed.relationship_label)
    _append(lines, "語氣距離", seed.tone_distance)
    _append(lines, "熟悉度邊界", seed.familiarity_boundary)
    _append(lines, "居住安排", seed.living_arrangement)
    _append(lines, "可知道的背景", seed.known_context)
    return lines


def _append(lines: list[str], label: str, value: str) -> None:
    text = (value or "").strip()
    if text:
        lines.append(f"- {label}：{text}")


def _schedule_policy_text(seed: CharacterOperatorRelationshipSeed) -> str:
    policy = seed.schedule_involvement_policy
    if policy == "mention_only":
        return "可準備話題或想起使用者偏好，但不可說成已約好共同活動。"
    if policy == "invite_required":
        return "可以安排想邀請使用者的活動，但必須說成邀請或詢問，不可當成已約定。"
    if policy == "shared_allowed":
        return "可以安排使用者明確允許的共同日常，但仍不可杜撰未提供的共同往事。"
    return "不把使用者排進角色行程。"

File: frontend/src/utils/api/initialRelationship.ts
import axios from 'axios'
import type { ScheduleInvolvementPolicy } from '@/types/character'

/**
 * 玩家編輯自己與角色的完整初始關係 seed（IR2）。
 *
 * 對應後端 `GET`/`PATCH /characters/{id}/initial-relationship`（IR1）：
 * PATCH 是 tri-state——欄位缺席＝不動，空字串＝清空，有值＝設定。
 * `confirmed_by_user` 不在可寫欄位裡：它只在建立當下由玩家確認一次，
 * 之後的編輯沒有資格重新聲稱那個時間點的確認。
 */

export interface InitialRelationshipSeed {
  character_id: string
  operator_id: string
  /** false＝這對玩家 × 角色還沒有任何 seed，其餘欄位都是空的預設值。 */
  has_seed: boolean
  relationship_label: string
  known_context: string
  living_arrangement: string
  user_address_name: string
  character_address_name: string
  tone_distance: string
  familiarity_boundary: string
  schedule_involvement_policy: ScheduleInvolvementPolicy
  proactive_permission: boolean
  proactive_cadence_hint: string
  user_profile_notes: string
  confirmed_by_user: boolean
}

/** PATCH payload：每個欄位缺席＝不動，空字串＝清空。由呼叫端只放「真的改過」的欄位。 */
export interface InitialRelationshipEditPatch {
  relationship_label?: string
  known_context?: string
  living_arrangement?: string
  user_address_name?: string
  character_address_name?: string
  tone_distance?: string
  familiarity_boundary?: string
  schedule_involvement_policy?: ScheduleInvolvementPolicy
  proactive_permission?: boolean
  proactive_cadence_hint?: string
  user_profile_notes?: string
}

const BASE = '/api/v1'

function url(characterId: string): string {
  return `${BASE}/characters/${encodeURIComponent(characterId)}/initial-relationship`
}

/** 讀出這對「玩家 × 角色」目前的整份關係 seed；沒設定過時回傳空白預設值（`has_seed: false`）。 */
export async function getInitialRelationship(
  characterId: string,
): Promise<InitialRelationshipSeed> {
  const { data } = await axios.get<InitialRelationshipSeed>(url(characterId))
  return data
}

/**
 * 只送真的改過的欄位（tri-state）。稱呼兩欄位也走這裡——後端會內部轉呼
 * `relationship-names` 的邏輯，rename-log 與已學到的人設投影一樣會同步，
 * 前端不需要另外呼叫 `/relationship-names`。
 */
export async function updateInitialRelationship(
  characterId: string,
  patch: InitialRelationshipEditPatch,
): Promise<InitialRelationshipSeed> {
  const { data } = await axios.patch<InitialRelationshipSeed>(url(characterId), patch)
  return data
}
